bson/codec_options.py

```diff
@@ -32,10 +33,82 @@ def _raw_document_class(document_class):
     return marker == _RAW_BSON_DOCUMENT_MARKER
 
 
+class TypeCodecBase(object):
+    """Base class for defining type codec classes which describe how a
+    custom type can be transformed to/from one of the types BSON already
+    understands, and can encode/decode.
+
+    Codec classes must implement the ``python_type`` property, and the
+    ``transform_python`` method to facilitate encoding, or the ``bson_type``
+    property and ``transform_bson`` method to facilitate decoding. A single
+    codec class may facilitate both encoding and decoding."""
+    @property
+    def python_type(self):
+        """The Python type to be converted into something serializable."""
+        raise NotImplementedError
+
+    @property
+    def bson_type(self):
+        """The BSON type to be converted into our own type."""
+        raise NotImplementedError
+
+    def transform_bson(self, value):
+        """Convert the given BSON value into our own type."""
+        raise NotImplementedError
+
+    def transform_python(self, value):
+        """Convert the given Python object into something serializable."""
+        raise NotImplementedError
+
+
+class TypeRegistry(object):
+    """Encapsulates type codecs used in encoding and / or decoding BSON.
+
+    ``TypeRegistry`` can be initialized with an arbitrary number of type
+    codecs::
+
+      >>> from bson.codec_options import TypeRegistry
+      >>> type_registry = TypeRegistry(Codec1, Codec2, Codec3, ...)
+
+    If multiple codecs try to transform a single python or BSON type,
+    the transformation described by the last type codec prevails."""
+    def __init__(self, *type_codecs):
+        self.__args = type_codecs
+        self._encoder_map = {}
+        self._decoder_map = {}
+        for codec in type_codecs:
+            if not isinstance(codec, TypeCodecBase):
+                raise TypeError(
+                    "Expected an instance of %s, got %r instead" % (
+                    TypeCodecBase.__name__, codec))
+            try:
+                python_type = codec.python_type
+            except NotImplementedError:
+                pass
+            else:
+                self._encoder_map[python_type] = codec.transform_python
```

Comment:
@ShaneHarvey @behackett do you think a warning should be generated when multiple codecs target the same type (and therefore, one of them gets overridden)? I have added a note about to the function docstring above, but if you think that is not enough, I can add a warning.

Reply:
Let's not add a warning for now.